### PR TITLE
Feat: add support for subenvironments

### PR DESCRIPTION
**Pull request by TomerHeber**
### Solution

Part of the effort to add support to subenvironments.
In this PR I added a helper function that extracts the subenvironments from the schema to an array of structs.
Unit test was added.

### CODE REVIEW

File: env0/resource_environment.go [hunk 1]
```diff
@@ -14,6 +14,47 @@ import (
...
+
+		if err := readResourceDataEx(prefix, &subEnvironment, d); err != nil {
+			return nil, err
+		}
+
+		configurationPrefix := prefix + ".configuration"
+		if configuration, ok := d.GetOk(configurationPrefix); ok {
+			subEnvironment.Configuration = getConfigurationVariablesFromSchema(configuration.([]interface{}))
+
+			for i := range subEnvironment.Configuration {
```
Comment (TomerHeber) [inline env0/resource_environment.go:47]: reviewing the API call in the UI I observed the scope is "ENVIRONMENT".

File: env0/resource_environment.go [hunk 2]
```diff
@@ -14,6 +14,47 @@ import (
...
+	numberOfSubEnvironments := len(isubEnvironments.([]interface{}))
+
+	var subEnvironments []SubEnvironment
+
+	for i := 0; i < numberOfSubEnvironments; i++ {
+		prefix := fmt.Sprintf("sub_environment_configuration.%d", i)
+
+		var subEnvironment SubEnvironment
+
+		if err := readResourceDataEx(prefix, &subEnvironment, d); err != nil {
```
Comment (TomerHeber) [inline env0/resource_environment.go:39]: the magic or "readResourceData".

File: env0/resource_environment.go [hunk 3]
```diff
@@ -14,6 +14,47 @@ import (
...
+			return nil, err
+		}
+
+		configurationPrefix := prefix + ".configuration"
+		if configuration, ok := d.GetOk(configurationPrefix); ok {
+			subEnvironment.Configuration = getConfigurationVariablesFromSchema(configuration.([]interface{}))
```
Comment (TomerHeber) [inline env0/resource_environment.go:45]: Configuration is not handled by readResourceData (for various reasons). Luckily there's already a function that does it for us.

File: env0/resource_environment.go [hunk 4]
```diff
@@ -237,6 +278,11 @@ func resourceEnvironment() *schema.Resource {
 				Optional:    true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
+						"id": {
```
Comment (TomerHeber) [inline env0/resource_environment.go:281]: this will be required later.

File: env0/utils_test.go [hunk 5]
```diff
@@ -338,3 +339,75 @@ func TestWriteResourceDataOmitEmpty(t *testing.T) {
...
+					Type:        (*client.ConfigurationVariableType)(intPtr(0)),
+					Schema: &client.ConfigurationVariableSchema{
+						Type: "string",
+					},
+				},
+			},
+		},
+	}
+
+	d := schema.TestResourceDataRaw(t, resourceEnvironment().Schema, map[string]interface{}{
```
Comment (TomerHeber) [inline env0/utils_test.go:383]: 2 subenvironments.
in the second subenironment I added 2 configurations.

**Review by eladmosh — APPROVED** [review]

**Pull request merged**